consensus/istanbul/core/handler_test.go

```diff
@@ -36,90 +38,77 @@ func TestHandleMsg(t *testing.T) {
 	v0 := sys.backends[0]
 	r0 := v0.engine.(*core)
 
-	m, _ := Encode(&istanbul.Subject{
+	m := istanbul.NewPrepareMessage(&istanbul.Subject{
 		View: &istanbul.View{
 			Sequence: big.NewInt(0),
 			Round:    big.NewInt(0),
 		},
 		Digest: common.BytesToHash([]byte("1234567890")),
-	})
+	}, v0.Address())
+
 	// with a matched payload. istanbul.MsgPreprepare should match with *istanbul.Preprepare in normal case.
-	msg := &istanbul.Message{
-		Code:      istanbul.MsgPreprepare,
-		Msg:       m,
-		Address:   v0.Address(),
-		Signature: []byte{},
-	}
+	m.Code = istanbul.MsgPreprepare
 
-	_, val := v0.Validators(nil).GetByAddress(v0.Address())
-	if err := r0.handleCheckedMsg(msg, val); err != errFailedDecodePreprepare {
-		t.Errorf("error mismatch: have %v, want %v", err, errFailedDecodePreprepare)
-	}
+	payload, err := m.Payload()
+	require.NoError(t, err)
 
-	m, _ = Encode(&istanbul.Preprepare{
+	err = r0.handleMsg(payload)
+	assert.Error(t, err)
```

Comment:
I'm a little worried that these tests are no longer testing what we want them to test.
Previously we were looking for the specific error (at least on the first 3 cases), but we send in the correct message (or we accidentally admit incorrect messages), this test will still pass (first is we need to sign, then make sure that the state is properly set up to accept the message based on it's code).

Reply:
Hey @trianglesphere, are you saying that we need to ensure that the mock engine needs to be setup correctly, otherwise practically any input would return an error from `handleCheckedMsg`?

Reply:
Yea. I think that the first set of message needs signatures and then the core to be in the correct state (such that they would accept the message if the payload was correct). If the core is not in the correct state for the message type, it rejects it with err old message or err future message.

 For the later tests with totally corrupt payloads, b/c we don't report a specific error, I think the best we can do is keep checking for an error.

Reply:
Looking at the previous tests I think what they were really trying to assert is that if a message is invalid then when decoded an error is returned.

Since all the decoding is now handled in `FromPayload`, rather than being spread out across the 3 message handlers ,I think we could replace these test cases with direct calls to `FromPayload`. We wouldn't be asserting that errors raised by `FromPayload` were propagated by `handleMsg`, but the previous tests called `handleCheckedMsg` directly and also didn't assert that errors raised by `handleCheckedMsg` were propagated by `handleMsg` (except for the malicious payload case, which hasn't really changed so we can keep it as it was). So I think this approach would be roughly equivalent.

Reply:
@trianglesphere I've now updated the tests in line with my previous comment.